crates/bgp-pkt/src/iana.rs

```diff
@@ -1354,6 +1354,396 @@ impl TryFrom<u8> for AigpAttributeType {
+pub enum BgpLsPrefixDescriptorTlvType {
+    MultiTopologyIdentifier = 263,
+    OspfRouteType = 264,
+    IpReachabilityInformation = 265,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
+pub struct UnknownPrefixDescriptorTlvType(pub u16);
+
+impl From<BgpLsPrefixDescriptorTlvType> for u16 {
+    fn from(value: BgpLsPrefixDescriptorTlvType) -> Self {
+        value as u16
+    }
+}
+
+impl TryFrom<u16> for BgpLsPrefixDescriptorTlvType {
+    type Error = UnknownPrefixDescriptorTlvType;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match Self::from_repr(value) {
+            Some(val) => Ok(val),
+            None => Err(UnknownPrefixDescriptorTlvType(value)),
+        }
+    }
+}
+
+#[repr(u16)]
+#[derive(Display, FromRepr, Copy, Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
+pub enum BgpLsLinkDescriptorTlvType {
+    LinkLocalRemoteIdentifiers = 258,
+    IPv4InterfaceAddress = 259,
+    IPv4NeighborAddress = 260,
+    IPv6InterfaceAddress = 261,
+    IPv6NeighborAddress = 262,
+    MultiTopologyIdentifier = 263,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
+pub struct UnknownLinkDescriptorTlvType(pub u16);
+
+impl From<BgpLsLinkDescriptorTlvType> for u16 {
+    fn from(value: BgpLsLinkDescriptorTlvType) -> Self {
+        value as u16
+    }
+}
+
+impl TryFrom<u16> for BgpLsLinkDescriptorTlvType {
+    type Error = UnknownLinkDescriptorTlvType;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match Self::from_repr(value) {
+            Some(val) => Ok(val),
+            None => Err(UnknownLinkDescriptorTlvType(value)),
+        }
+    }
+}
+
+#[repr(u16)]
+#[derive(Display, FromRepr, Copy, Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
+pub enum BgpLsDescriptorTlvType {
+    LocalNodeDescriptor = 256,
+    RemoteNodeDescriptor = 257,
+    MultiTopologyIdentifier = 263,
+    OspfRouteType = 264,
+    IpReachabilityInformation = 265,
+    LinkLocalRemoteIdentifiers = 258,
+    IPv4InterfaceAddress = 259,
+    IPv4NeighborAddress = 260,
+    IPv6InterfaceAddress = 261,
+    IPv6NeighborAddress = 262,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
+pub struct UnknownDescriptorTlvType(pub u16);
+
+impl From<BgpLsDescriptorTlvType> for u16 {
+    fn from(value: BgpLsDescriptorTlvType) -> Self {
+        value as u16
+    }
+}
+
+impl TryFrom<u16> for BgpLsDescriptorTlvType {
+    type Error = UnknownDescriptorTlvType;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match Self::from_repr(value) {
+            Some(val) => Ok(val),
+            None => Err(UnknownDescriptorTlvType(value)),
+        }
+    }
+}
+
+/// Aggregate of [BgpLsLinkAttribute] [BgpLsNodeAttribute]
+/// [BgpLsPrefixAttribute]
+#[repr(u16)]
+#[derive(Display, FromRepr, Copy, Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
+pub enum BgpLsAttributeTlv {
```

Comment:
Maybe better to redefine this as: (to avoid duplicate code copying)
```rust
pub enum BgpLsAttributeTlv {
  BgpLisLinkAttribute(BgpLisLinkAttribute),
  BgpLsNodeAttribute(BgpLsNodeAttribute),
   BgpLsPrefixAttribute(BgpLsPrefixAttribute),
}
```

Reply:
some values (LocalNodeIpv4RouterId = 1028 for example) are in both `BgpLsLinkAttribute` and `BgpLsNodeAttribute` when matching i will not know which variant (Link/Node) this value corresponds to

Reply:
I assume you mean in the deserialization part ..if so, you can use the `alt` nom combinator with a `map`, check https://github.com/rust-bakery/nom/blob/main/doc/choosing_a_combinator.md#choice-combinators

Also `BgpLsAttributeTlv` is repeated twice with the same name, I would pick different names for bgp_ls and iana to make it more clear and less confusing.

Reply:
yes i mean in the deserialization. i understand the alt combinator with map. However, then, when i parse the tlv type 1028 i still have no way to know whether it should be a `BgpLsLinkAttribute::LocalNodeIpv4RouterId ` or `BgpLsLinkAttribute::LocalNodeIpv4RouterId`. this will probably depend on the order i put the parsers in alt for which it will be arbitrary. I did not find something in rfc7752 saying that the BGP LS Attribute must be before the MP_REACH containing the NLRI so I can't assume that I will know whether the NLRI is Link or Node.